### PR TITLE
Fix for persistince numpy arrays of object dtype

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Previously, when a numpy array could not be stored directly, it was attempted to json-serialize the object. However, this didn't work. The new solution now recursively calls `get_state` on the objects inside the array.

To check the effect, the testing had to be fixed to be able to compare generic objects (à la `Foo() == Foo()`). This in turn unmasked some existing bugs where the dtype might change after a save-load roundtrip (e.g. `[OneHotEncoder()] - AssertionError: assert dtype('<U6') == object`).

It also reveals that for `SGDClassifier` and `SGDOneClassSVM`, there actually seem to be incorrectly stored values, which looks like a bug in our current implementation.

Locally, test failures went from 15 to 13.

Ready for review @adrinjalali

### CODE REVIEW

**Review by adrinjalali — APPROVED** [review]
Thanks, would need to start testing our equality thing at some point I think lol

File: skops/io/_numpy.py [hunk 1]
```diff
@@ -24,13 +23,14 @@ def ndarray_get_state(obj, dst):
...
-        # object arrays cannot be saved with allow_pickle=False, therefore we
-        # convert them to a list and store them as a json file.
-        f_name = f"{uuid4()}.json"
-        with open(Path(dst) / f_name, "w") as f:
-            f.write(json.dumps(obj.tolist()))
+        # Object arrays cannot be saved with allow_pickle=False, therefore we
+        # convert them to a list and recursively call get_state on it.
+        if obj.dtype == object:
+            obj = get_state(obj.tolist(), dst)
+            res["content"] = obj["content"]
```
Comment (adrinjalali) [inline skops/io/_numpy.py:30]: Oh wait, how large are these, you sure we should put them in the schema.json?

**Comment by adrinjalali** [comment]
I'll merge and we can think about splitting the schema.json maybe later.

File: skops/io/_numpy.py (hunk 1 above)
Comment (BenjaminBossan) [reply skops/io/_numpy.py:30]: Since we call `get_state`, if there are any numpy arrays contained, they'll be put on disk, so we should be good, right?

**Comment by BenjaminBossan** [comment]
> Thanks, would need to start testing our equality thing at some point I think lol

Agree

**Pull request merged**